sdk/helper/docker/testhelpers.go

```diff
@@ -148,38 +179,112 @@ func (d *Runner) StartService(ctx context.Context, connect ServiceAdapter) (*Ser
 	return serv, err
 }
 
+type LogConsumerWriter struct {
+	consumer func(string)
+}
+
+func (l LogConsumerWriter) Write(p []byte) (n int, err error) {
+	// TODO this assumes that we're never passed partial log lines, which
+	// seems a safe assumption for now based on how docker looks to implement
+	// logging, but might change in the future.
+	scanner := bufio.NewScanner(bytes.NewReader(p))
+	scanner.Buffer(make([]byte, 64*1024), bufio.MaxScanTokenSize)
+	for scanner.Scan() {
+		l.consumer(scanner.Text())
+	}
+	return len(p), nil
+}
+
+var _ io.Writer = &LogConsumerWriter{}
+
 // StartNewService will start the runner's configured docker container but with the
 // ability to control adding a name suffix or forcing a local address to be returned.
 // 'addSuffix' will add a random UUID to the end of the container name.
 // 'forceLocalAddr' will force the container address returned to be in the
 // form of '127.0.0.1:1234' where 1234 is the mapped container port.
 func (d *Runner) StartNewService(ctx context.Context, addSuffix, forceLocalAddr bool, connect ServiceAdapter) (*Service, string, error) {
-	container, hostIPs, containerID, err := d.Start(context.Background(), addSuffix, forceLocalAddr)
+	if d.RunOptions.PreDelete {
+		name := d.RunOptions.ContainerName
+		matches, err := d.DockerAPI.ContainerList(ctx, types.ContainerListOptions{
```

Comment:
Am I correct in understanding this is exclusive of `t.Parallel()`? If we specify `addSuffix` to allow tests to execute in parallel, here we'd kill all variants? 

A docs follow-up should be sufficient if so.

Reply:
I don't think so?  It seems to me we're using exact match rather than prefix match in the filter.

The main use case for the pre-delete is for local testing where we're not using addSuffix, to prevent a prior failed run from interfering.  We should probably do that labels TODO so that in local testing with addSuffix, old containers get cleaned up; they won't interfere, but they'll take up memory.

Reply:
Ah, I missed the exact match. I think this is good. Yeah, I'm curious what you would put in the label though... perhaps the test name?